# **STAGE 2 - ETL**

## **Objectives**

In this Jupyter Notebook, I will be following the ETL process on the raw dataset: `bank-churners-anonymised.csv`

The process will be as follows:
* **Extract**: I will load the dataset into a pandas DataFrame and perform an initial inspection of its structure, data types and quality.
* **Transform**: I will clean and standardise the dataset, change data types where necessary, remove any null values, and engineer new features.
* **Load**: Finally, I will export the cleaned DataFrame as a new CSV file, ready to be used in the next Jupyter Notebook - **Stage 3 - Visualisation**.

## **Inputs**

* The data input I need for this section is an anonymised dataset created during **Stage 1 - Anonymise**
* It is found in my project here: `datasets/anonymised-raw-data/bank-churners-anonymised.csv`.

## **Outputs**

* At the end of this notebook, I created a new .csv file which has been cleaned and is ready for analysis.
* The relative path for this .csv file is: `datasets/cleaned-data/bank-churners-cleaned.csv`

---

## **Change Working Directory**

* We are assuming you will store the notebooks in a subfolder, therefore when running the notebook in the editor, you will need to change the working directory

We need to change the working directory from its current folder to its parent folder
* We access the current directory with os.getcwd()

In [1]:
import os
current_dir = os.getcwd()
current_dir

'/Users/elliebrawn/Documents/vscode-projects/credit-card-customer-churn-analysis/jupyter_notebooks'

We want to make the parent of the current directory the new current directory
* os.path.dirname() gets the parent directory
* os.chir() defines the new current directory

In [2]:
os.chdir(os.path.dirname(current_dir))
print("You set a new current directory")

You set a new current directory


Confirm the new current directory

In [3]:
current_dir = os.getcwd()
current_dir

'/Users/elliebrawn/Documents/vscode-projects/credit-card-customer-churn-analysis'

---

## **Import Packages and Libraries**

In [4]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
sns.set_style("whitegrid")
sns.set_palette("PiYG", n_colors=6)

---

## **Core Statistical Concepts**

Before diving into this project, I am first going to outline some key statistical concepts that are foundational to data analysis. Because many of these will appear in my project, it is important to outline what they mean and how they are used.

I'll split these up into three main sections: 

### **Descriptive Statistics**
Descriptive statistics summarise features of a data set. They can describe things like the **central tendency** and the **variability**.

| **Concept** | **Sub-Concept** | **Description** |
| ---------- | --------------| -------- |
| **Central Tendency** | - | The central tendency refers to a value that shows the middle or typical point in a dataset. |
| - | **Mean** | The mean is the average value of a dataset. <br><br>***How to Calculate***: Add all the values together and divide by the total number of observations. <br><br>***Pandas method to calculate***: Use `.mean()`. <br><br> The mean provides a simple measure of the central tendency of the data, although it can be strongly affected by unusually large or small values (outliers). |
| - | **Median** | The median is the middle value - it can be thought of as the element that divides the data into two equal parts. <br><br>***How to Calculate***: Arrange the data in ascending or descending order. If the number of elements is odd, the median is the middle element in the arranged values. If the number of observations is even, the median is given by the mean of the two middle elements. <br><br>***Pandas method to calculate***: Use `.median()`. <br><br>Unlike the mean, the median is less affected by outliers, making it useful for datasets with skewed distributions. |
| - | **Mode** | The mode is the value that appears most often (has the highest frequency) in a dataset. <br><br>***How to Calculate***: The easiest way is to order the numbers in ascending or descending order. Then, count the number of times a value appears. <br><br>***Pandas method to Calculate***: Use `.mode()`. <br><br>A dataset can be **Uni-modal** (only one mode), **Bi-modal** (two modes), or **Multi-modal** (more than two modes). |
| **Variability** | - | Measures of variability are also known as dispersion or spread and show how spread out the data is. Some of the ways to measure this are **standard deviation**, **range** and **variance**. |
| - | **Standard Deviation** | Standard deviation measures how spread out the observations are around the mean. It is used by data analysts to assess the distribution of the data. A small standard deviation means values tend to be close to the mean. A large standard deviation indicates greater variability. |

### **Inferential Statistics**
Whereas descriptive statistics summarise the data you have, inferential statistics use data to draw conclusions beyond the observed sample.

| **Concept** | **Sub-Concept** | **Description** |
| ---------- | --------------| -------- |
| **Hypothesis Testing** | - | This is a statistical method used to determine whether there is sufficient evidence to support a claim. <br><br> A **null hypothesis (H0)** represents no effect or difference, and an **alternative hypothesis (Ha or H1)**, which represents an effect or difference. A statistical test and a significance level are then used to determine whether the evidence is strong enough to reject the null hypothesis. |

### **Probability**
Statistical probability measures how likely an event is to occur and is expressed as a number between 0 and 1.

| **Concept** | **Sub-Concept** | **Description** |
| ---------- | --------------| -------- |
| **p-value** | - | The p-value is a statistical measure and specific probability that tells you the likelihood that an observed outcome is the result of chance or that it could have occured under the null hypothesis. It tells you whether a result is **statistically significant**. |

### **Why are these core concepts so important in Data Analysis?**
All of the concepts listed together are the foundation of data analyis. They help to transform raw data into meaningful information and help to support evidence-based decision-making.

> *Notes on Process*:
<br><br>This section has been copied over from the last Code Institute project I submitted, so the write-up will be the same.
<br><br>

---

## **1. Extraction**

In this first section, Extraction, I am looking to gain a baseline understanding of what my data looks like before I move on to cleaning and transforming.

I am using the acronym - **DISH** - to guide this process, where the letters stand for:

| Letter | Function / Attribute |
| ------ | ------- |
|**D**| `.describe`|
|**I**| `.info()`|
|**S**| `.shape`|
|**H**| `.head()`|

I will also use the following extraction steps:

| Function / Attribute |
| -----|
|`.dtypes`|
|`.isnull.sum()`|

### **1.1 Load the Raw Dataset**

Firstly, I will load the raw dataset - `bank-churners-anonymised.csv` - that I am going to be working on for this ETL Jupyter Notebook.

In [5]:
# Load the raw dataset
df_anonymised = pd.read_csv("datasets/anonymised-raw-data/bank-churners-anonymised.csv")

# Display the first few rows of the dataset to make sure the data has loaded correctly
df_anonymised.head()

,anonymised_clientnum,Attrition_Flag,Customer_Age,Gender,Dependent_count,Education_Level,Marital_Status,Income_Category,Card_Category,Months_on_book,...,Months_Inactive_12_mon,Contacts_Count_12_mon,Credit_Limit,Total_Revolving_Bal,Avg_Open_To_Buy,Total_Amt_Chng_Q4_Q1,Total_Trans_Amt,Total_Trans_Ct,Total_Ct_Chng_Q4_Q1,Avg_Utilization_Ratio
0,022e52f0a251431f954db620fd0c87ac3c523c60cc5980...,Existing Customer,45,M,3,High School,Married,$60K - $80K,Blue,39,...,1,3,12691.0,777,11914.0,1.335,1144,42,1.625,0.061
1,2731de4ed9ecb2a3ab828448bfda6137e5c7571e1f7576...,Existing Customer,49,F,5,Graduate,Single,Less than $40K,Blue,44,...,1,2,8256.0,864,7392.0,1.541,1291,33,3.714,0.105
2,75dac624c2bbdb15b16abc0350820bcf598c012a76b102...,Existing Customer,51,M,3,Graduate,Married,$80K - $120K,Blue,36,...,1,0,3418.0,0,3418.0,2.594,1887,20,2.333,0.000
3,1aaada0cd1f7dc23d83a171beec9f398cfac1471841843...,Existing Customer,40,F,4,High School,Unknown,Less than $40K,Blue,34,...,4,1,3313.0,2517,796.0,1.405,1171,20,2.333,0.760
4,1811f55b3210153f77e41ceceb61cc181d2edc4e65947e...,Existing Customer,40,M,3,Uneducated,Married,$60K - $80K,Blue,21,...,1,0,4716.0,0,4716.0,2.175,816,28,2.500,0.000


### **1.2 Display the Shape of the Dataset**

In [6]:
# Display the shape of the dataset
print(f"Shape of the dataset: {df_anonymised.shape}")

Shape of the dataset: (10127, 21)


The dataset has 10,127 rows and 21 columns.

### **1.3 Display the Data Types for each Column**

In order to identify any data type mismatches, I am going to inspect the data types for each column of my DataFrame.

In [7]:
# Outline the data types of each column in the dataset
print(f"Data types of each column:\n\n{df_anonymised.dtypes}")

Data types of each column:

anonymised_clientnum         object
Attrition_Flag               object
Customer_Age                  int64
Gender                       object
Dependent_count               int64
Education_Level              object
Marital_Status               object
Income_Category              object
Card_Category                object
Months_on_book                int64
Total_Relationship_Count      int64
Months_Inactive_12_mon        int64
Contacts_Count_12_mon         int64
Credit_Limit                float64
Total_Revolving_Bal           int64
Avg_Open_To_Buy             float64
Total_Amt_Chng_Q4_Q1        float64
Total_Trans_Amt               int64
Total_Trans_Ct                int64
Total_Ct_Chng_Q4_Q1         float64
Avg_Utilization_Ratio       float64
dtype: object


The only columns I want to change here are `Total_Trans_Amt` and `Total_Revolving_Bal`: these are both listed as integers, but because they refer to a currency amount, I want to store them as floats. I will do this in my **Transform** section. 

### **1.4 Display a Summary of all Columns**

In [8]:
# Use .info() to get a concise summary of the dataset, including the number of non-null entries
df_anonymised.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10127 entries, 0 to 10126
Data columns (total 21 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   anonymised_clientnum      10127 non-null  object 
 1   Attrition_Flag            10127 non-null  object 
 2   Customer_Age              10127 non-null  int64  
 3   Gender                    10127 non-null  object 
 4   Dependent_count           10127 non-null  int64  
 5   Education_Level           10127 non-null  object 
 6   Marital_Status            10127 non-null  object 
 7   Income_Category           10127 non-null  object 
 8   Card_Category             10127 non-null  object 
 9   Months_on_book            10127 non-null  int64  
 10  Total_Relationship_Count  10127 non-null  int64  
 11  Months_Inactive_12_mon    10127 non-null  int64  
 12  Contacts_Count_12_mon     10127 non-null  int64  
 13  Credit_Limit              10127 non-null  float64
 14  Total_

This method builds on what we have already established using `.shape` and `.dtypes`. 

It shows us that we have 10,127 entries and 21 data columns and the data types for each of these columns. 

Additionally, we can see an insight into `Non-Null Count`, aka the number of values in each column that aren't null. 
* I will explore this further down in **1.6 Check for Missing Values in the Dataset**. 
* This summary of the dataset appears to show that there are no missing values in this dataset.

### **1.5 Display Summary Statistics for the Dataset**

In [9]:
# Use .describe() for summary statistics of the numerical columns in the dataset
df_anonymised.describe()

,Customer_Age,Dependent_count,Months_on_book,Total_Relationship_Count,Months_Inactive_12_mon,Contacts_Count_12_mon,Credit_Limit,Total_Revolving_Bal,Avg_Open_To_Buy,Total_Amt_Chng_Q4_Q1,Total_Trans_Amt,Total_Trans_Ct,Total_Ct_Chng_Q4_Q1,Avg_Utilization_Ratio
count,10127.000000,10127.000000,10127.000000,10127.000000,10127.000000,10127.000000,10127.000000,10127.000000,10127.000000,10127.000000,10127.000000,10127.000000,10127.000000,10127.000000
mean,46.325960,2.346203,35.928409,3.812580,2.341167,2.455317,8631.953698,1162.814061,7469.139637,0.759941,4404.086304,64.858695,0.712222,0.274894
std,8.016814,1.298908,7.986416,1.554408,1.010622,1.106225,9088.776650,814.987335,9090.685324,0.219207,3397.129254,23.472570,0.238086,0.275691
min,26.000000,0.000000,13.000000,1.000000,0.000000,0.000000,1438.300000,0.000000,3.000000,0.000000,510.000000,10.000000,0.000000,0.000000
25%,41.000000,1.000000,31.000000,3.000000,2.000000,2.000000,2555.000000,359.000000,1324.500000,0.631000,2155.500000,45.000000,0.582000,0.023000
50%,46.000000,2.000000,36.000000,4.000000,2.000000,2.000000,4549.000000,1276.000000,3474.000000,0.736000,3899.000000,67.000000,0.702000,0.176000
75%,52.000000,3.000000,40.000000,5.000000,3.000000,3.000000,11067.500000,1784.000000,9859.000000,0.859000,4741.000000,81.000000,0.818000,0.503000
max,73.000000,5.000000,56.000000,6.000000,6.000000,6.000000,34516.000000,2517.000000,34516.000000,3.397000,18484.000000,139.000000,3.714000,0.999000


As you can see, `.describe()` shows some of the core descriptive statistics measures that I have described in more detail in **Core Statistical Concepts**. 

There are no minus numbers included in any of the numerical columns and with a quick look over all of the figures, nothing seems particularly anomalous. I will look at the interquartile range and outlier detection in **2. Transform** if any outliers need handling.

### **1.6 Check for Missing Values**

In [10]:
## Check for null values in the dataset
print(f"Number of null values in each column:\n\n{df_anonymised.isnull().sum()}")

Number of null values in each column:

anonymised_clientnum        0
Attrition_Flag              0
Customer_Age                0
Gender                      0
Dependent_count             0
Education_Level             0
Marital_Status              0
Income_Category             0
Card_Category               0
Months_on_book              0
Total_Relationship_Count    0
Months_Inactive_12_mon      0
Contacts_Count_12_mon       0
Credit_Limit                0
Total_Revolving_Bal         0
Avg_Open_To_Buy             0
Total_Amt_Chng_Q4_Q1        0
Total_Trans_Amt             0
Total_Trans_Ct              0
Total_Ct_Chng_Q4_Q1         0
Avg_Utilization_Ratio       0
dtype: int64


As was indicated when using `.describe()`, there are no missing values that we will need to address.

---

## **2. Transform**

In this next section, Transform, my aim is to convert my raw and extracted dataset into a clean, well-structured dataset that is feature-rich to support the hypotheses I will investigate in the **Stage 3 - Visualisation** Jupyter Notebook.

I will create a copy of my `df_anonymised` DataFrame so as not to transform the raw dataset.

In [11]:
#Create a copy of the DataFrame
df_etl = df_anonymised.copy()

### **2.1 Change Data Types**

As identified above, I want to change `Total_Trans_Amt` and `Total_Revolving_Bal` from integers to floats to reflect the fact that they relate to currency.

In [12]:
# Change the data type of the Total_Revolving_Bal column to float
df_etl["Total_Revolving_Bal"] = df_etl["Total_Revolving_Bal"].astype("float64").round(2)

# Change the data type of the Total_Trans_Amt column to float
df_etl["Total_Trans_Amt"] = df_etl["Total_Trans_Amt"].astype("float64").round(2)

# Confirm these data type changes have been made
print(f"Data types of each column after conversion:\n\n{df_etl.dtypes}")

Data types of each column after conversion:

anonymised_clientnum         object
Attrition_Flag               object
Customer_Age                  int64
Gender                       object
Dependent_count               int64
Education_Level              object
Marital_Status               object
Income_Category              object
Card_Category                object
Months_on_book                int64
Total_Relationship_Count      int64
Months_Inactive_12_mon        int64
Contacts_Count_12_mon         int64
Credit_Limit                float64
Total_Revolving_Bal         float64
Avg_Open_To_Buy             float64
Total_Amt_Chng_Q4_Q1        float64
Total_Trans_Amt             float64
Total_Trans_Ct                int64
Total_Ct_Chng_Q4_Q1         float64
Avg_Utilization_Ratio       float64
dtype: object
